# Saloni Patel (A4 - 4.9 - Part 1 Intro to Data Visualisation with Python)

# Importing Libraries and Data

### 1 - Importing Libraries

In [4]:
# importing libraries

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

### 2 - Defining Path

In [5]:
# Creating Path for ease of code

path = r'C:\Users\psalo\OneDrive\Study\CF - DA Immersion\A4 - Python Fundamentals for Data Analyst\A4 - Instacart Basket Analysis 07-2025'


### 3 - Importing 'ords_prods_merge_post_4_8' from Prepared Data Folder

There were no changes made to ords_prods_merge in 4.9 so the most uptodate version of the df is from 4.8

In [6]:
# Importing ords_prods_merge_post_4_8 from Prepared Data Folder.

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_post_4_8.pkl'))
print(f"Number of rows imported: {ords_prods_merge.shape[0]}")

Number of rows imported: 32404859


In [27]:
ords_prods_merge.describe()

product_id      aisle_id  department_id        prices      order_id  \
count  3.240486e+07  3.240486e+07   3.240486e+07  3.239973e+07  3.240486e+07   
mean   2.559866e+04  7.119612e+01   9.919792e+00  7.790994e+00  1.710745e+06   
std    1.408400e+04  3.821139e+01   6.281485e+00  4.241809e+00  9.872988e+05   
min    1.000000e+00  1.000000e+00   1.000000e+00  1.000000e+00  2.000000e+00   
25%    1.354400e+04  3.100000e+01   4.000000e+00  4.200000e+00  8.559470e+05   
50%    2.530200e+04  8.300000e+01   9.000000e+00  7.400000e+00  1.711049e+06   
75%    3.794700e+04  1.070000e+02   1.600000e+01  1.130000e+01  2.565499e+06   
max    4.968800e+04  1.340000e+02   2.100000e+01  2.500000e+01  3.421083e+06   

            user_id  order_number  orders_day_of_week  order_hour_of_day  \
count  3.240486e+07  3.240486e+07        3.240486e+07       3.240486e+07   
mean   1.029372e+05  1.714230e+01        2.738867e+00       1.342515e+01   
std    5.946610e+04  1.753532e+01        2.090077e+00       4.246380e+00   
min    1.000000e+00  1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.142200e+04  5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.026160e+05  1.100000e+01        3.000000e+00       1.300000e+01   
75%    1.543890e+05  2.400000e+01        5.000000e+00       1.600000e+01   
max    2.062090e+05  9.900000e+01        6.000000e+00       2.300000e+01   

       days_since_last_order  add_to_cart_order     reordered     max_order  \
count           3.240486e+07       3.240486e+07  3.240486e+07  3.240486e+07   
mean            1.084114e+01       8.352547e+00  5.895873e-01  3.305217e+01   
std             8.552435e+00       7.127071e+00  4.919087e-01  2.515525e+01   
min             0.000000e+00       1.000000e+00  0.000000e+00  1.000000e+00   
25%             5.000000e+00       3.000000e+00  0.000000e+00  1.300000e+01   
50%             7.000000e+00       6.000000e+00  1.000000e+00  2.600000e+01   
75%             1.400000e+01       1.100000e+01  1.000000e+00  4.700000e+01   
max             3.000000e+01       1.450000e+02  1.000000e+00  9.900000e+01   

       avg_prod_price  median_days_betn_orders  
count    3.240486e+07             3.240486e+07  
mean     7.790996e+00             9.749833e+00  
std      7.348183e-01             6.357435e+00  
min      1.000000e+00             0.000000e+00  
25%      7.378488e+00             6.000000e+00  
50%      7.811940e+00             7.000000e+00  
75%      8.229328e+00             1.200000e+01  
max      2.320000e+01             3.000000e+01

In [28]:
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 25 columns):
 #   Column                   Dtype   
---  ------                   -----   
 0   product_id               int64   
 1   product_name             object  
 2   aisle_id                 int64   
 3   department_id            int64   
 4   prices                   float64 
 5   order_id                 int64   
 6   user_id                  int64   
 7   eval_set                 object  
 8   order_number             int64   
 9   orders_day_of_week       int64   
 10  order_hour_of_day        int64   
 11  days_since_last_order    float64 
 12  add_to_cart_order        int64   
 13  reordered                int64   
 14  _merge                   category
 15  price_range_loc          object  
 16  busiest_day              object  
 17  busiest_days             object  
 18  busiest_period_of_day    object  
 19  max_order                int64   
 20  loyalty_flag          

### 4 - Importing customers.csv from Original Data Folder

In [9]:
# Importing customers.csv from Original Data Folder.

df_cust = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers', 'customers.csv'))
print(f"Number of rows imported: {df_cust.shape[0]}")

Number of rows imported: 206209


### 5 - Checking the imported df_cust data shape

In [10]:
# Checking the shape

df_cust.shape

(206209, 10)

### 6 - Checking the imported df_cust data types

In [11]:
# Checking the Types

df_cust.dtypes

user_id          int64
First Name      object
Surnam          object
Gender          object
STATE           object
Age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

### 7 - Checking the imported df_cust data head

In [12]:
# Checking the Head

df_cust.head(5)

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

### 8 - Checking the imported df_cust data for statistical analysis

In [16]:
# Checking the statistical data

df_cust.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

#### Statistical analysis >>> There are no outliers in the columns that are relevant - age - dependents - income

### 9 - Renaming the columns in df_cust

Making sure to add _ instead of space in variable titles and also ensuring that they are all in lower cap for consistency

In [18]:
df_cust = df_cust.rename(columns={ 
    'First Name' : 'first_name',
    'Surnam' : 'last_name', 
    'STATE' : 'state', 
    'Age' : 'age', 
    'Gender' : 'gender', 
    'n_dependants' : 'dependents', 
    'fam_status' : 'marital_status'})

In [19]:
# Checking to see if the change has been in effect

df_cust.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   dependents marital_status  income  
0           3        married  165665  
1           0         single   59285  
2           2        married   99568  
3           0         single   42049  
4           1        married   40374

#### Renaming the variable >>>> Success

#### NOTE >>>> There are no redundant columns apparent in this dataset so there is no need to remove anything so far 

### 10 - Checking for duplicates in df_cust

In [26]:
# Exploring for duplicate rows

df_cust[df_cust.duplicated()]


Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, dependents, marital_status, income]
Index: []

#### Duplicate rows >>>> There are no duplicate rows in the data frame

### 11 - Checking for missing values in df_cust

In [25]:
# exploring for missing data

df_cust.isnull().sum()


user_id               0
first_name        11259
last_name             0
gender                0
state                 0
age                   0
date_joined           0
dependents            0
marital_status        0
income                0
dtype: int64

#### Missing Data >>>> There are 11,259 rows where first name is missing. However this should not be a constraint.

### 12 - Merging df_cust with ords_prods_merge using common user_id column. 

Types in both database are the same for this column

In [33]:
# merging using inner join 

df_ord_prod_user = pd.merge(ords_prods_merge, df_cust, on='user_id', how='inner')

### 13 - Checks on the merged data frame - df_ord_prod_user

In [43]:
# Checking the df_ord_prod_user - shape

df_ord_prod_user.shape

(32404859, 34)

In [44]:
# Checking the df_ord_prod_user - info

df_ord_prod_user.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 34 columns):
 #   Column                   Dtype   
---  ------                   -----   
 0   product_id               int64   
 1   product_name             object  
 2   aisle_id                 int64   
 3   department_id            int64   
 4   prices                   float64 
 5   order_id                 int64   
 6   user_id                  int64   
 7   eval_set                 object  
 8   order_number             int64   
 9   orders_day_of_week       int64   
 10  order_hour_of_day        int64   
 11  days_since_last_order    float64 
 12  add_to_cart_order        int64   
 13  reordered                int64   
 14  _merge                   category
 15  price_range_loc          object  
 16  busiest_day              object  
 17  busiest_days             object  
 18  busiest_period_of_day    object  
 19  max_order                int64   
 20  loyalty_flag          

In [45]:
# Checking the df_ord_prod_user - describe

df_ord_prod_user.describe()

product_id      aisle_id  department_id        prices      order_id  \
count  3.240486e+07  3.240486e+07   3.240486e+07  3.239973e+07  3.240486e+07   
mean   2.559866e+04  7.119612e+01   9.919792e+00  7.790994e+00  1.710745e+06   
std    1.408400e+04  3.821139e+01   6.281485e+00  4.241809e+00  9.872988e+05   
min    1.000000e+00  1.000000e+00   1.000000e+00  1.000000e+00  2.000000e+00   
25%    1.354400e+04  3.100000e+01   4.000000e+00  4.200000e+00  8.559470e+05   
50%    2.530200e+04  8.300000e+01   9.000000e+00  7.400000e+00  1.711049e+06   
75%    3.794700e+04  1.070000e+02   1.600000e+01  1.130000e+01  2.565499e+06   
max    4.968800e+04  1.340000e+02   2.100000e+01  2.500000e+01  3.421083e+06   

            user_id  order_number  orders_day_of_week  order_hour_of_day  \
count  3.240486e+07  3.240486e+07        3.240486e+07       3.240486e+07   
mean   1.029372e+05  1.714230e+01        2.738867e+00       1.342515e+01   
std    5.946610e+04  1.753532e+01        2.090077e+00       4.246380e+00   
min    1.000000e+00  1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.142200e+04  5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.026160e+05  1.100000e+01        3.000000e+00       1.300000e+01   
75%    1.543890e+05  2.400000e+01        5.000000e+00       1.600000e+01   
max    2.062090e+05  9.900000e+01        6.000000e+00       2.300000e+01   

       days_since_last_order  add_to_cart_order     reordered     max_order  \
count           3.240486e+07       3.240486e+07  3.240486e+07  3.240486e+07   
mean            1.084114e+01       8.352547e+00  5.895873e-01  3.305217e+01   
std             8.552435e+00       7.127071e+00  4.919087e-01  2.515525e+01   
min             0.000000e+00       1.000000e+00  0.000000e+00  1.000000e+00   
25%             5.000000e+00       3.000000e+00  0.000000e+00  1.300000e+01   
50%             7.000000e+00       6.000000e+00  1.000000e+00  2.600000e+01   
75%             1.400000e+01       1.100000e+01  1.000000e+00  4.700000e+01   
max             3.000000e+01       1.450000e+02  1.000000e+00  9.900000e+01   

       avg_prod_price  median_days_betn_orders           age    dependents  \
count    3.240486e+07             3.240486e+07  3.240486e+07  3.240486e+07   
mean     7.790996e+00             9.749833e+00  4.946527e+01  1.501896e+00   
std      7.348183e-01             6.357435e+00  1.848558e+01  1.118865e+00   
min      1.000000e+00             0.000000e+00  1.800000e+01  0.000000e+00   
25%      7.378488e+00             6.000000e+00  3.300000e+01  1.000000e+00   
50%      7.811940e+00             7.000000e+00  4.900000e+01  2.000000e+00   
75%      8.229328e+00             1.200000e+01  6.500000e+01  3.000000e+00   
max      2.320000e+01             3.000000e+01  8.100000e+01  3.000000e+00   

             income  
count  3.240486e+07  
mean   9.943773e+04  
std    4.305727e+04  
min    2.590300e+04  
25%    6.700400e+04  
50%    9.661800e+04  
75%    1.279120e+05  
max    5.939010e+05

In [41]:
# Checking the df_ord_prod_user - head

df_ord_prod_user.head(5)

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id eval_set  order_number  orders_day_of_week  ...  \
0   3139998      138    prior            28                   6  ...   
1   1977647      138    prior            30                   6  ...   
2    389851      709    prior             2                   0  ...   
3    652770      764    prior             1                   3  ...   
4   1813452      764    prior             3                   4  ...   

   order_frequency_flag  first_name  last_name  gender      state age  \
0     Frequent customer     Charles        Cox    Male  Minnesota  81   
1     Frequent customer     Charles        Cox    Male  Minnesota  81   
2     Frequent customer     Deborah      Glass  Female    Vermont  66   
3     Frequent customer     Heather      Myers  Female  Wisconsin  40   
4     Frequent customer     Heather      Myers  Female  Wisconsin  40   

  date_joined dependents marital_status  income  
0    8/1/2019          1        married   49620  
1    8/1/2019          1        married   49620  
2   6/16/2018          2        married  158302  
3    2/9/2020          3        married   31308  
4    2/9/2020          3        married   31308  

[5 rows x 34 columns]

### 14 - Export of the updated version

### Exporting

In [46]:
# Exporting final df in pkl version and printing the number of rows exported for one final verification

num_rows = len(df_ord_prod_user)  
print(f"{num_rows} rows exported to pkl file.")

df_ord_prod_user.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'df_ord_prod_user.pkl'))

32404859 rows exported to pkl file.
